File: lib/wType/address.cpp
#include "address.h"

W::Address::Address():
    Object(),
    data(new List())
{

}

W::Address::Address(const W::Address::InitList& list):
    Object(),
    data(new List())
{
    InitList::const_iterator itr = list.begin();
    InitList::const_iterator end = list.end();
    
    for (; itr != end; ++itr)
    {
        data->emplace_back(String(*itr));
    }
}

W::Address::Address(const W::Address& original):
    Object(),
    data(new List(*original.data))
{

}

W::Address::~Address()
{
    delete data;
}

W::Address& W::Address::operator=(const W::Address& original)
{
    if (&original != this) 
    {
        data->clear();
        data->insert(data->end(), original.data->begin(), original.data->end());
    }
    return *this;
}

W::Object::size_type W::Address::length() const
{
    return data->size();
}

W::Object::objectType W::Address::getType() const
{
    return type;
}

W::Object::StdStr W::Address::toString() const
{
    StdStr str;
    
    List::const_iterator itr;
    List::const_iterator beg = data->begin();
    List::const_iterator end = data->end();
    
    str += '[';
    for (itr = beg; itr != end; ++itr)
    {
        if (itr != beg)
        {
            str += ", ";
        }
        str += itr->toString();
    }
    str += ']';
    
    return str;
}

W::Object* W::Address::copy() const
{
    return new Address(*this);
}

void W::Address::serialize(W::ByteArray& out) const
{
    out.push32(length());
    
    List::const_iterator itr;
    List::const_iterator beg = data->begin();
    List::const_iterator end = data->end();
    
    for (itr = beg; itr != end; ++itr)
    {
        itr->serialize(out);
    }
}

void W::Address::deserialize(W::ByteArray& in)
{
    data->clear();
    
    size_type length = in.pop32();
    
    for (size_type i = 0; i != length; ++i)
    {
        data->emplace_back(String());
        data->back().deserialize(in);
    }
}

bool W::Address::operator<(const W::Address& other) const
{
    return *data < *other.data;
}

bool W::Address::operator>(const W::Address& other) const
{
    return *data > *other.data;
}

bool W::Address::operator==(const W::Address& other) const
{
    return *data == *other.data;
}

bool W::Address::operator!=(const W::Address& other) const
{
    return *data != *other.data;
}

bool W::Address::operator<=(const W::Address& other) const
{
    return *data <= *other.data;
}

bool W::Address::operator>=(const W::Address& other) const
{
    return *data >= *other.data;
}

bool W::Address::begins(const W::Address& other) const
{
    if (other.length() > length())
    {
        return false;
    }
    
    List::const_iterator itr_o = other.data->begin();
    List::const_iterator end_o = other.data->end();
    
    List::const_iterator itr_i = data->begin();
    
    while (itr_o != end_o)
    {
        if (*itr_o != *itr_i) {
            return false;
        }
        
        ++itr_o;
        ++itr_i;
    }
    
    return true;
}

bool W::Address::ends(const W::Address& other) const
{
    if (other.length() > length())
    {
        return false;
    }
    
    List::const_reverse_iterator itr_o = other.data->rbegin();
    List::const_reverse_iterator end_o = other.data->rend();
    
    List::const_reverse_iterator itr_i = data->rbegin();
    
    while (itr_o != end_o)
    {
        if (*itr_o != *itr_i) {
            return false;
        }
        
        ++itr_o;
        ++itr_i;
    }
    
    return true;
}

bool W::Address::contains(const W::Address& other, int position) const
{
    if (other.length() > length() - position)
    {
        return false;
    }
    
    bool res = true;
    
    List::const_iterator itr_o = other.data->begin();
    List::const_iterator end_o = other.data->end();
    
    List::const_iterator itr_i = data->begin();
    
    for (int i = 0; i < position; ++i) {
        ++itr_i;
    }
    
    while (res == true && itr_o != end_o)
    {
        res = *itr_o == *itr_i;
        
        ++itr_o;
        ++itr_i;
    }
    
    return res;
}

W::Address& W::Address::operator+=(const W::Address& other)
{
    data->insert(data->end(), other.data->begin(), other.data->end());
    return *this;
}

W::Address& W::Address::operator+=(const W::String& other)
{
    data->push_back(other);
    return *this;
}

W::Address& W::Address::operator+=(const W::String::u16string& other)
{
    String hop(other);
    return operator+=(hop);
}

W::Address W::Address::operator>>(W::Object::size_type count) const
{
    W::Address res;
    if (count < length())
    { 
        List::const_iterator itr = data->end();
        for (size_type i = 0; i != count; ++i)
        {
            itr--;
        }
        List::const_iterator beg = data->begin();
        res.data->insert(res.data->end(), beg, itr);
    }
    return res;
}

W::Address W::Address::operator<<(W::Object::size_type count) const
{
    W::Address res;
    if (count < length())
    { 
        List::const_iterator itr = data->begin();
        for (size_type i = 0; i != count; ++i)
        {
            itr++;
        }
        List::const_iterator end = data->end();
        res.data->insert(res.data->end(), itr, end);
    }
    return res;
}

W::Address W::Address::operator+(const W::Address& other) const
{
    W::Address res;
    res += *this;
    res += other;
    
    return res;
}

W::Address & W::Address::operator+=(const Uint64& other)
{
    data->push_back(String(other.toString()));
    
    return *this;
}


W::Address W::Address::operator+(const Uint64& other) const
{
    W::Address res;
    res += *this;
    res += other;
    
    return res;
}


const W::String& W::Address::front() const
{
    return data->front();
}

const W::String& W::Address::back() const
{
    return data->back();
}

bool W::Address::operator==(const W::Object& other) const
{
    if (sameType(other)) {
        return operator==(static_cast<const W::Address&>(other));
    } else {
        return false;
    }
}

W::Object::size_type W::Address::size() const
{
    size_type size = 4;
    List::const_iterator itr = data->begin();
    List::const_iterator end = data->end();
    
    for (; itr != end; ++itr)
    {
        size += itr->size();
    }
    
    return size;
}

File: libjs/wType/address.js
"use strict";
var Object = require("./object");
var String = require("./string");

var Address = Object.inherit({
    "className": "Address",
    "constructor": function(data) {
        Object.fn.constructor.call(this);
        
        this._data = [];
        this._parseSource(data || []);
    },
    "destructor": function() {
        this.clear();
        
        Object.fn.destructor.call(this);
    },
    "<": function(other) {
        if (!(other instanceof Address)) {
            throw new Error("Can compare Address only with Address");
        }
        var hopMe;
        var hopOt;
        
        for (var i = 0; i < this._data.length; ++i) {
            hopMe = this._data[i];
            hopOt = other._data[i];
            
            if (hopOt === undefined) {
                return false;
            }
            
            if (hopMe["<"](hopOt)) {
                return true;
            }
            if (hopMe[">"](hopOt)) {
                return false;
            }
        }
        return this._data.length < other._data.length;
    },
    ">": function(other) {
        if (!(other instanceof Address)) {
            throw new Error("Can compare Address only with Address");
        }
        var hopMe;
        var hopOt;
        
        for (var i = 0; i < this._data.length; ++i) {
            hopMe = this._data[i];
            hopOt = other._data[i];
            
            if (hopOt === undefined) {
                return true;
            }
            
            if (hopMe[">"](hopOt)) {
                return true;
            }
            if (hopMe["<"](hopOt)) {
                return false;
            }
        }
        return this._data.length > other._data.length;
    },
    "==": function(other) {
        if (this.getType() !== other.getType()) {
            return false;
        }
        
        if (this._data.length !== other._data.length) {
            return false;
        }
        
        var hopMe;
        var hopOt;
        
        for (var i = 0; i < this._data.length; ++i) {
            hopMe = this._data[i];
            hopOt = other._data[i];
            if ( !(hopMe["=="](hopOt)) ) {
                return false;
            }
        }
        return true;
    },
    "+": function(other) {
        var res = this.clone();
        res["+="](other);
        return res;
    },
    "+=": function(other) {
        if (other instanceof Address) {
            for (var i = 0; i < other._data.length; ++i) {
                this._data.push(other._data[i].clone());
            }
        } else {
            throw new Error("Can add to Address only Address");
        }
        
        return this;
    },
    "<<": function(n) {
        var res = new Address();
        for (var i = n; i < this._data.length; ++i) {
            res._data.push(this._data[i].clone());
        }
        return res;
    },
    ">>": function(n) {
        var res = new Address();
        for (var i = 0; i < this._data.length - n; ++i) {
            res._data.push(this._data[i].clone());
        }
        return res;
    },
    "clear": function() {
        for (var i = 0; i < this._data.length; ++i) {
            this._data[i].destructor();
        }
        this._data = [];
    },
    "clone": function() {
        var clone = new Address();
        
        for (var i = 0; i < this._data.length; ++i) {
            clone._data.push(this._data[i].clone());
        }
        
        return clone;
    },
    "deserialize": function(ba) {
        this.clear()
        var length = ba.pop32();
        
        for (var i = 0; i < length; ++i) {
            var hop = new String();
            hop.deserialize(ba);
            this._data.push(hop);
        }
    },
    "serialize": function(ba) {
        ba.push32(this._data.length)
        
        for (var i = 0; i < this._data.length; ++i) {
            this._data[i].serialize(ba);
        }
    },
    "length": function() {
        return this._data.length;
    },
    "size": function() {
        var size = 4;
        for (var i = 0; i < this._data.length; ++i) {
            size += this._data[i].size();
        }
        
        return size;
    },
    "toString": function() {
        var str = "";
        str += "["
        
        for (var i = 0; i < this._data.length; ++i) {
            if (i !== 0) {
                str +=", ";
            }
            str += this._data[i].toString();
        }
        
        str += "]";
        return str;
    },
    "begins": function(other) {
        var size = other._data.length;
        if (size > this._data.length) {
            return false;
        }
        
        for (var i = 0; i < size; ++i) {
            var myHop = this._data[i];
            var othHop = other._data[i];
            
            if (!myHop["=="](othHop)) {
                return false;
            }
        }
        
        return true;
    },
    "ends": function(other) {
        var size = other._data.length;
        if (size > this._data.length) {
            return false;
        }
        
        for (var i = 1; i <= size; ++i) {
            var myHop = this._data[this._data.length - i];
            var othHop = other._data[other._data.length - i];
            
            if (!myHop["=="](othHop)) {
                return false;
            }
        }
        
        return true;
    },
    "back": function() {
        return this._data[this._data.length - 1].clone();
    }, 
    "front": function() {
        return this._data[0].clone();
    },
    "toArray": function() {
        var arr = [];
        for (var i = 0; i < this._data.length; ++i) {
            arr.push(this._data[i].toString());
        }
        return arr;
    },
    "_parseSource": function(data) {
        for (var i = 0; i < data.length; ++i) {
            this._data.push(new String(data[i]));
        }
    }
});

module.exports = Address;
